mantis-server/mantis-server-agent/src/main/java/io/mantisrx/server/agent/TaskExecutor.java

```diff
@@ -416,25 +421,31 @@ public CompletableFuture<Ack> submitTask(ExecuteStageRequest request) {
         WrappedExecuteStageRequest wrappedRequest =
             new WrappedExecuteStageRequest(PublishSubject.create(), request);
 
-        Task task = new Task(
-            wrappedRequest,
-            workerConfiguration,
-            masterMonitor,
-            classLoaderHandle,
-            subscriptionStateHandlerFactory,
-            Optional.of(getHostname()),
-            Optional.empty());
-
-        setCurrentTask(task);
-
-        scheduleRunAsync(this::startCurrentTask, 0, TimeUnit.MILLISECONDS);
-        return CompletableFuture.completedFuture(Ack.getInstance());
+        try {
+            UserCodeClassLoader userCodeClassLoader = ClassLoaderHandle.createUserCodeClassloader(request, classLoaderHandle);
+            ClassLoader cl = userCodeClassLoader.asClassLoader();
+            final ITask task = InstantiationUtil.instantiate("io.mantisrx.server.worker.Task", ITask.class, cl);
```

Comment:
What do you think about using ServiceLoader::load (https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html#load-java.lang.Class-java.lang.ClassLoader-) instead of this? This would help us avoid hardcoding the task class name and have that specified as part of the runtime. It's cleaner, IMO, as you can easily migrate to TaskV2 in the future by just changing the runtime. 

One more suggestion: instead of using setters for passing the context to the ITask instance, how about we create an ITask::Factory abstraction, which takes these parameters as context and returns an ITask instance as a return parameter. This would result in the runtime having to define an ITask::Factory implementation (say TaskFactory) as part of META-INF/services.

Reply:
Yes, we can try using ServiceLoader. That would be cleaner.

I'm not sure what introducing an extra layer of Factory buy us. Can you explain more? If the concern is called multiple setXXX methods, we could replace with a single init(...) method in the ITask interface. When we introduce ITaskV2, we would also need ITaskV2.Factory. I want to understand better what's the benefits of the added layer.